File: Courses/PHP Web Basics - май 2019/16.ExerciseEncapsulationInheritance/01.05.PizzaCalories/Pizza.php
<?php


class Pizza
{
    /**
     * @var string
     */
    private $name;
    /**
     * @var int
     */
    private $numberOfToppings;
    /**
     * @var Dough $dough
     */
    private $dough;

    /**
     * @var Topping[] $toppings
     */
    private $toppings = [];

    /**
     * Pizza constructor.
     * @param string $name
     * @param int $numberOfToppings
     * @throws Exception
     */
    public function __construct(string $name, int $numberOfToppings)
    {
        $this->setName($name);
        $this->setNumberOfToppings($numberOfToppings);

    }

    /**
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @param string $name
     * @throws Exception
     */
    public function setName(string $name): void
    {
        if (strlen($name) < 1 || strlen($name) > 15) {
            throw new Exception("Pizza name should be between 1 and 15 symbols.");
        }
        $this->name = $name;
    }

    /**
     * @param Dough $dough
     */


    /**
     * @return void $topping
     */

    public function addTopping(Topping $topping) : void
    {
        $this->toppings[] = $topping;
    }

    /**
     * @param $numberOfToppings
     * @throws Exception
     */
    public function setNumberOfToppings($numberOfToppings)
    {
        if ($numberOfToppings < 0 || $numberOfToppings > 10) {
            throw new \Exception("Number of toppings should be in range [0..10].");
        }
        $this->numberOfToppings = $numberOfToppings;
    }
    public function setDough(Dough $dough): void
    {
        $this->dough = $dough;
    }
    public function getTotalCalories() :float {
        $calories = $this->dough->getCalories();

            foreach ($this->toppings as $topping) {
                $calories += $topping->getCalories();
            }
            return $calories;
    }

    public function __toString()
    {
        $totalCalories = number_format($this->getTotalCalories(), 2, ".","");
        $this->name[0] = strtoupper($this->name[0]);
        return "{$this->name} - {$totalCalories} Calories.";
    }


}

File: Courses/C# OOP/Basics/08. Encapsulation Exercise/08. Encapsulation Exercise/05.PizzaCalories/Pizza.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PizzaCalories
{
    public class Pizza
    {
        private string name;
        private Dough dough;
        private List<Topping> toppings;

        public Pizza(string name)
        {
            Name = name;
            Toppings = new List<Topping>();
        }

        public string Name
        {
            get => name;
            set
            {
                if (value.Length < 1 || value.Length > 15)
                {
                    throw new ArgumentException("Pizza name should be between 1 and 15 symbols.");
                }
                name = value;
            }
        }
        public Dough Dough { get => dough; set => dough = value; }
        public List<Topping> Toppings
        {
            get => toppings; set
            {
                toppings = value;
            }
        }

        public void AddTopping(Topping topping)
        {
            if (this.Toppings.Count() == 10)
            {
                throw new ArgumentException("Number of toppings should be in range [0..10].");
            }
            toppings.Add(topping);
        }
        private double GetCalories()
        {
            double doughCal = Dough.DoughCalories;
            double toppingCal = Toppings.Sum(c => c.ToppingCalories);
            return doughCal + toppingCal;
        }
        public override string ToString()
        {
            return $"{Name} - {GetCalories():F2} Calories.";
        }
    }
}
